zerver/lib/outgoing_webhook.py

```diff
@@ -0,0 +1,91 @@
+
+    if recipient_type_name == 'stream':
+        recipients = [message['display_recipient']]
+        check_send_message(bot_user, get_client("OutgoingWebhookResponse"), recipient_type_name, recipients,
+                           message['subject'], response_message_content, realm, forwarder_user_profile=bot_user)
+    else:
+        # Private message; only send if the bot is there in the recipients
+        recipients = [recipient['email'] for recipient in message['display_recipient']]
+        if bot_user.email in recipients:
+            check_send_message(bot_user, get_client("OutgoingWebhookResponse"), recipient_type_name, recipients,
+                               message['subject'], response_message_content, realm, forwarder_user_profile=bot_user)
+
+def succeed_with_message(event, success_message):
+    # type: (Dict[str, Any], Text) -> None
+    send_response_message(event['user_profile_id'], event['message'], success_message)
+
+def fail_with_message(event, failure_message):
+    # type: (Dict[str, Any], Text) -> None
+    send_response_message(event['user_profile_id'], event['message'], failure_message)
+
+def request_retry(event, failure_message):
+    # type: (Dict[str, Any], Text) -> None
+    event['retry'] += 1
+    if event['retry'] > 3:
+        bot_user = get_user_profile_by_id(event['user_profile_id'])
+        fail_with_message(event, failure_message)
+        logging.warning("Maximum retries exceeded for trigger:%s event:%s" % (bot_user.email, event['command']))
+    else:
+        queue_json_publish("outgoing_webhooks", event, lambda x: None)
+
+def do_rest_call(rest_operation, event, timeout=None):
+    # type: (Dict[str, Any], Dict[str, Any], Any) -> None
+    rest_operation_validator = check_dict([
+        ('method', check_string),
+        ('relative_url_path', check_string),
+        ('request_kwargs', check_dict([])),
+        ('base_url', check_string),
+    ])
+
+    error = rest_operation_validator('rest_operation', rest_operation)
+    if error:
+        raise JsonableError(_("%s") % (error,))
+
+    http_method = rest_operation['method']
+    final_url = urllib.parse.urljoin(rest_operation['base_url'], rest_operation['relative_url_path'])
+    request_kwargs = rest_operation['request_kwargs']
+    request_kwargs['timeout'] = timeout
+
+    try:
+        response = requests.request(http_method, final_url, data=json.dumps(event), **request_kwargs)
+        if str(response.status_code).startswith('2'):
+            succeed_with_message(event, "Success!")
+
+        # On 50x errors, try retry
+        elif str(response.status_code).startswith('5'):
+            request_retry(event, 'Maximum retries exceeded')
+        else:
+            fail_with_message(event, "Failure!")
+
+    except requests.exceptions.Timeout:
+        logging.info("Trigger event %s on %s timed out. Retrying" % (event["command"], event['service_name']))
+        request_retry(event, 'Maximum retries exceeded')
```

Comment:
Style nit: Would it be possible to move the message `Maximum retries exceeded` into the `request_retry` function?  Code that calls `request_retry` shouldn't have to worry about this error message.